File: 395.longest-substring-with-at-least-k-repeating-characters.java
/*
 * @lc app=leetcode id=395 lang=java
 *
 * [395] Longest Substring with At Least K Repeating Characters
 */

// @lc code=start
class Solution {
    public int longestSubstring(String s, int k) {
        char[] str = s.toCharArray();
        int[] countMap = new int[26];
        int maxUnique = getMaxUniqueLetters(s);

        int result = 0;
        for (int currUnique = 1; currUnique <= maxUnique; currUnique++) {
            // reset countMap
            Arrays.fill(countMap, 0);

            int windowStart = 0, windowEnd = 0, idx = 0, unique = 0, countAtLeastK = 0;
            while (windowEnd < str.length) {
                // If the number of unique characters in the sliding window
                // is less than or equal to currUnique,
                // expand the sliding window
                if (unique <= currUnique) {
                    idx = str[windowEnd] - 'a';
                    if (countMap[idx] == 0)
                        unique++;
                    countMap[idx]++;
                    if (countMap[idx] == k)
                        countAtLeastK++;
                    windowEnd++;
                }
                // shrink the sliding window
                else {
                    idx = str[windowStart] - 'a';
                    if (countMap[idx] == k)
                        countAtLeastK--;
                    countMap[idx]--;
                    if (countMap[idx] == 0)
                        unique--;
                    windowStart++;
                }

                // the number of unique characters in the sliding window
                // is equal to currUnique,
                // and all of unique characters are counted at least k times
                if (unique == currUnique && unique == countAtLeastK)
                    result = Math.max(windowEnd - windowStart, result);
            }
        }

        return result;
    }

    // get the maximum number of unique letters in the string s
    int getMaxUniqueLetters(String s) {
        boolean map[] = new boolean[26];
        int maxUnique = 0;
        for (int i = 0; i < s.length(); i++) {
            if (!map[s.charAt(i) - 'a']) {
                maxUnique++;
                map[s.charAt(i) - 'a'] = true;
            }
        }
        return maxUnique;
    }

    // public int longestSubstring(String s, int k) {
    // int[] count = new int[26];
    // for (int i = 0; i < s.length(); i++)
    // count[s.charAt(i) - 'a']++;

    // for (int i = 0; i < s.length(); i++) {
    // if (count[s.charAt(i) - 'a'] < k) {
    // int j = i + 1;
    // while (j < s.length() && count[s.charAt(j) - 'a'] < k)
    // j++;
    // return Math.max(longestSubstring(s.substring(0, i), k),
    // longestSubstring(s.substring(j), k));
    // }
    // }

    // return s.length();
    // }
}
// @lc code=end

File: 395.longest-substring-with-at-least-k-repeating-characters.py
#
# @lc app=leetcode id=395 lang=python3
#
# [395] Longest Substring with At Least K Repeating Characters
#

# @lc code=start


class Solution:
    def longestSubstring(self, s: str, k: int) -> int:
        # If k is 0, every substring is valid
        if k == 0:
            return len(s)

        n = len(s)
        res = 0
        # Maximum distinct characters we need to consider is limited by the string itself
        max_unique = len(set(s))

        # Try every possible target number of distinct characters in the window
        for target in range(1, max_unique + 1):
            count = defaultdict(int)
            left = 0
            right = 0
            unique = 0  # current number of distinct characters in window
            no_less_than_k = 0  # number of characters in window with freq >= k

            # Expand window by moving right
            while right < n:
                c = s[right]
                if count[c] == 0:
                    unique += 1
                count[c] += 1
                if count[c] == k:
                    no_less_than_k += 1
                right += 1

                # Shrink window from the left if we exceed target distinct chars
                while unique > target:
                    d = s[left]
                    if count[d] == k:
                        no_less_than_k -= 1
                    count[d] -= 1
                    if count[d] == 0:
                        unique -= 1
                    left += 1

                # If window has exactly 'target' distinct chars,
                # and all of them appear at least k times,
                # update the maximum length
                if unique == target and no_less_than_k == target:
                    res = max(res, right - left)

        return res


# @lc code=end
